src/devices/Arduino/ArduinoGpioControllerDriver.cs

```diff
@@ -0,0 +1,305 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Device.Gpio;
+using System.Linq;
+using System.Threading;
+
+namespace Iot.Device.Arduino
+{
+    internal class ArduinoGpioControllerDriver : GpioDriver
+    {
+        private readonly ArduinoBoard _arduinoBoard;
+        private readonly List<SupportedPinConfiguration> _supportedPinConfigurations;
+        private readonly Dictionary<int, CallbackContainer> _callbackContainers;
+        private readonly ConcurrentDictionary<int, PinMode> _pinModes;
+        private readonly object _callbackContainersLock;
+        private readonly AutoResetEvent _waitForEventResetEvent;
+
+        internal ArduinoGpioControllerDriver(ArduinoBoard arduinoBoard, List<SupportedPinConfiguration> supportedPinConfigurations)
+        {
+            _arduinoBoard = arduinoBoard ?? throw new ArgumentNullException(nameof(arduinoBoard));
+            _supportedPinConfigurations = supportedPinConfigurations ?? throw new ArgumentNullException(nameof(supportedPinConfigurations));
+            _callbackContainers = new Dictionary<int, CallbackContainer>();
+            _waitForEventResetEvent = new AutoResetEvent(false);
+            _callbackContainersLock = new object();
+            _pinModes = new ConcurrentDictionary<int, PinMode>();
+
+            PinCount = _supportedPinConfigurations.Count;
+            _arduinoBoard.Firmata.DigitalPortValueUpdated += FirmataOnDigitalPortValueUpdated;
+        }
+
+        protected override int PinCount { get; }
+
+        /// <summary>
+        /// Arduino does not distinguish between logical and physical numbers, so this always returns identity
+        /// </summary>
+        protected override int ConvertPinNumberToLogicalNumberingScheme(int pinNumber)
+        {
+            return pinNumber;
+        }
+
+        protected override void OpenPin(int pinNumber)
+        {
+        }
+
+        protected override void ClosePin(int pinNumber)
+        {
+            _pinModes.TryRemove(pinNumber, out _);
```

Comment:
you're removing something on ClosePin but do nothing on OpenPin, that sounds like some inconsistency but didn't dig

Reply:
The Add() is on `SetPinMode`, which is always called after `Open()`.